pkg/querier/frontend2/frontend_scheduler_worker.go

```diff
@@ -0,0 +1,303 @@
+package frontend2
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/weaveworks/common/httpgrpc"
+	"google.golang.org/grpc"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+type frontendSchedulerWorkers struct {
+	services.Service
+
+	cfg             Config
+	log             log.Logger
+	frontendAddress string
+
+	// Channel with requests that should be forwarded to the scheduler.
+	requestsCh <-chan *frontendRequest
+
+	watcher services.Service
+
+	mu sync.Mutex
+	// Set to nil when stop is called... no more workers are created afterwards.
+	workers map[string]*frontendSchedulerWorker
+}
+
+func newFrontendSchedulerWorkers(cfg Config, frontendAddress string, requestsCh <-chan *frontendRequest, log log.Logger) (*frontendSchedulerWorkers, error) {
+	f := &frontendSchedulerWorkers{
+		cfg:             cfg,
+		log:             log,
+		frontendAddress: frontendAddress,
+		requestsCh:      requestsCh,
+		workers:         map[string]*frontendSchedulerWorker{},
+	}
+
+	w, err := NewDNSWatcher(cfg.SchedulerAddr, cfg.DNSLookupPeriod, f)
+	if err != nil {
+		return nil, err
+	}
+
+	f.watcher = w
+	f.Service = services.NewIdleService(f.starting, f.stopping)
+	return f, nil
+}
+
+func (f *frontendSchedulerWorkers) starting(ctx context.Context) error {
+	return services.StartAndAwaitRunning(ctx, f.watcher)
+}
+
+func (f *frontendSchedulerWorkers) stopping(_ error) error {
+	err := services.StopAndAwaitTerminated(context.Background(), f.watcher)
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	for _, w := range f.workers {
+		w.stop()
+	}
+	f.workers = nil
+
+	return err
+}
+
+func (f *frontendSchedulerWorkers) AddressAdded(address string) {
+	f.mu.Lock()
+	ws := f.workers
+	w := f.workers[address]
+	f.mu.Unlock()
+
+	// Already stopped or we already have worker for this address.
+	if ws == nil || w != nil {
+		return
+	}
+
+	level.Debug(f.log).Log("msg", "adding connection to scheduler", "addr", address)
```

Comment:
I left it as Debug because that's what `pkg/querier/frontend/worker.go` does. I agree it's useful, and I'm fine with using Info here, but then I would suggest changing `pkg/querier/frontend/worker.go` too. WDYT?

Reply:
Let's do both in a separate PR.